Add database validation rules

girardinsamuel opened the pull request:
I finally did not follow `not_taken` name as it could be used for a lot of checks (not specifically an email or a name for a User as the name could imply). I choose `exists_in_db` as `exists` is taken for dicts.

## Features:
- [x] add `exists_in_db` validation rule (need to be rewritten cleanly though) => Implements #22 
- [x] integrate it with new Masonite ORM
- [x] add more tests
- [x] add `unique_in_db` validation rule

## Wontfix
add "string" validator usage (https://docs.masoniteproject.com/advanced/validation#string-validation)
```python
errors = request.validate({
  'email': 'required|exists_in_db:users',
}) 
```
I did not add full string usage, as there are multiple arguments to parse and I guess it shouldn't be done in the rule but in the core validation code

## Usage
```python
validate.exists_in_db("email", "users")
validate.exists_in_db("name", "users", "first_name")
validate.exists_in_db("email", "app.models.User")
validate.exists_in_db("email", "app.models.User", connection="sqlite")
```
```python
validate.unique_in_db("email", "users")
validate.unique_in_db("my_unique_email", "app.User", "email")
validate.unique_in_db("email", "app.models.User")
```
Fixes #22

girardinsamuel committed:
add prototype of exists_in_db with Orator ORM

josephmancuso commented:
is there anymore thought on how to do this? Should we somehow put this in the ORMProvider? to add the validations to masonite ORM?

girardinsamuel commented:
Not really, but it's a really important feature for me. I was planning to resume my work on it now ORM is ready.

I guess it might requires some new helpers in ORM and in some way to be able to trigger these through validation class.
It would be great to be able to resolve models in the validation class to make the checks.

girardinsamuel commented on src/masonite/validation/Validator.py:
```diff
@@ -1029,6 +1029,46 @@ def negated_message(self, attribute):
+        # retrieving db connection will totally change with new ORM
+        # in addition, below is not really satisfying
+        from masonite.helpers import config
+        self.connection = config("database.db")
```
@josephmancuso 
(even if it has not yet be upgraded to ORM)
I was doing this: fetch the connection from the config and then prepare a query based on the validator parameters

josephmancuso replied:
get its not upgraded. To make this work you would have to do:

```
self.connection = config("database.db").get_query_builder()
```

Then you can make queries

```python
self.connection = config("database.db").get_query_builder()
self.connection.table(table).where(..)
```

josephmancuso commented on src/masonite/validation/Validator.py:
```diff
@@ -1029,6 +1029,46 @@ def negated_message(self, attribute):
+        # in addition, below is not really satisfying
+        from masonite.helpers import config
+        self.connection = config("database.db")
+        self.column = column
```
I actually like how you did this so I guess we can keep doing down this right

girardinsamuel replied:
okay cool ! 
I will upgrade this to use masonite ORM and see if I can work it out for the different use cases.

maybe I should take care to ensure that validation can works even if ORM is not there or installed

josephmancuso replied:
Yeah that's an important step. This package should be useable outside of even needing Masonite .. Having some methods where Masonite is required is the best of both world but shouldn't need a package to use this package

girardinsamuel committed:
Merge branch '3.0' into feat/22

girardinsamuel committed:
update test project to Masonite 3.0

girardinsamuel committed:
upgrade rule to masonite orm and fix test

girardinsamuel committed:
add more test

girardinsamuel committed:
update tests cmd with tag to not running tests requiring db on ci

girardinsamuel committed:
add unique validation rule

girardinsamuel committed:
try solution 1 to have a test project 2 and3 compatible

girardinsamuel commented:
@josephmancuso  I completed the implementation of the database rules 😉 

About the compatibility 2/3 and the use masonite ORM, here is a summary of my thoughts and what I needed to do:

The compatibility issue arise at different points:
- in the code of the new rules ✅  (imports are wrapped into the rules so no errors if not called)
- in the code of the tests ✅  (i put some pytest skip to skip the tests if masonite ORM was not installed)
- but independently of the two points above, I have seen if I installed masonite 3 and the orm in my env to run the tests (to test the databases rules), it won't work because the example project delivered with validation package will cause the tests to fail (that's the same issue I got the other day). Because of that, I adopted one not very elegant solution : make this test project compatible for both 2 and 3 (with the help for try statements and ifs ...). 

=> Now we can run the tests in the two envs (Masonite 2 AND Masonite 3 + ORM). If the 3 env, if Masonite ORM is not installed tests will run and database tests will be skipped.

Last point, because of the possibility to specify the db connection in the rule, I needed to add a test for it. This test requires that another connection would exists. As the default one was sqlite (for easing running tests), the new one must requires a db server (mysql, postgres ...). To ease running tests quickly and also through github actions, I added a pytest mark `db_tests` to be able to run all tests without this one. 
- on CI: tests without the tests marked `db_tests` are run
- locally: we can choose (I added a new command in makefile)


We could also go for other solutions, for now that's the one I followed. 
- we could install masonite 3 and orm in github workflow to run new db rules tests
- we could install a db server in github workflow as for orm, to be able to also the test marked with `db_tests` 
- we could have a workflow for masonite 2 and one for masonite 3 on CI
- ...

girardinsamuel committed:
update env example file

josephmancuso commented:
> About the compatibility 2/3 and the use masonite ORM

Remind me why we are supporting 2 and 3 in the same version of this package again? Why can't we just release version 4 or something if that makes it easier? 

> but independently of the two points above, I have seen if I installed masonite 3 and the orm in my env to run the tests (to test the databases rules), it won't work because the example project delivered with validation package will cause the tests to fail (that's the same issue I got the other day). Because of that, I adopted one not very elegant solution : make this test project compatible for both 2 and 3 (with the help for try statements and ifs ...).

yeah not a fan of that

> Last point, because of the possibility to specify the db connection in the rule, I needed to add a test for it. This test requires that another connection would exists. As the default one was sqlite (for easing running tests), the new one must requires a db server (mysql, postgres ...). To ease running tests quickly and also through github actions, I added a pytest mark db_tests to be able to run all tests without this one.

You can have another database connection that is also sqlite

girardinsamuel commented:
@josephmancuso 
> Remind me why we are supporting 2 and 3 in the same version of this package again? Why can't we just release version 4 or something if that makes it easier?

I guess it's just historical, no separation has been made yet because there was no point. We can actually release this in the 4 version which will target Masonite 3. And I guess for using this package outside Masonite you will use the 4 version.
So then I can remove all if/try about masonite version and it will be cleaner/easier to maintain

> You can have another database connection that is also sqlite

Oh ok, I guess you mean that I can name the connection differently but still be based on the sqlite driver. that will be handy for the test, I'll add this 👍

girardinsamuel committed:
revert pytest markers and use other sqlite connection for testing

girardinsamuel committed:
try to add env key to fix tests

girardinsamuel commented:
@josephmancuso we should decide how supporting Masonite 2 and Masonite 3 for this package.
This PR is ready, but it might have to be merged in a branch specific to masonite 3 version if you want that I remove the if statements for 2/3 compatibility 🤔 

Also in the requirements file there is a strange requirement for masonite, I guess it should be open for 2 and 3 versions ?
```
masonite>=2.3,<2.4.0
```

Commit:
prepare for 4.0 release

girardinsamuel committed:
Merge branch '3.0' into feat/22

girardinsamuel committed:
revert publish workflow

girardinsamuel committed:
fix typos in workflows

girardinsamuel committed:
add missing users migrations for tests

girardinsamuel committed:
fix tests

girardinsamuel committed:
switch to pendulum instead of pytz not installed anymore with masonite 3

girardinsamuel committed:
drop python 3.5 support as masonite dropped it too

girardinsamuel committed:
rename unique to unique_in_db

girardinsamuel commented:
@josephmancuso I made all the changes required to work with Masonite 3. CI is passing.
I defined setup.py version as 4.0.0. I guess now we have to merge this in 4.0 branch and release it ?
The docs update is ready here: https://github.com/MasoniteFramework/docs/pull/154

Do you want me to prepare something else ?

girardinsamuel commented:
I am closing this PR as it won't be integrated on this package but into Masonite core.
